tests/integration/pilot/trafficshifting/helper.go

```diff
@@ -17,71 +17,96 @@ package trafficshifting
 import (
 	"fmt"
 	"math"
-	"net/http"
 	"strings"
 	"testing"
+	"time"
 
 	"istio.io/istio/pkg/test/framework/components/apps"
 )
 
-func sendTraffic(t *testing.T, batchSize int, from apps.KubeApp, to string, hosts []string, weight []int32, errorBand float64) {
-
-	stop := make(chan bool, 1)
-	totalRequests := 0
-	hostnameHitCount := map[string]int{}
-	errorFailures := map[string]int{}
-
-	go func() {
-		for totalRequests < batchSize {
-			headers := http.Header{}
-			headers.Add("Host", to)
-			// This is a hack to remain infrastructure agnostic when running these tests
-			// We actually call the host set above not the endpoint we pass
-			resp, err := from.Call(from.EndpointForPort(80), apps.AppCallOptions{
-				Protocol: apps.AppProtocolHTTP,
-				Headers:  headers,
-			})
-			totalRequests++
-			for _, r := range resp {
-				for _, h := range hosts {
-					if strings.HasPrefix(r.Hostname, h+"-") {
-						hostnameHitCount[h]++
-						break
+func sendTraffic(t *testing.T, duration time.Duration, batchSize int, from apps.KubeApp, to apps.KubeApp, hosts []string, weight []int32, errorBand float64) {
+	const totalThreads = 10
+	var hitCount [totalThreads]map[string]int
+	var errorFailures [totalThreads]map[string]int
+
+	for i := 0; i < totalThreads; i++ {
+		hitCount[i] = map[string]int{}
+		errorFailures[i] = map[string]int{}
+
+		go func(id int) {
+			for {
+				resp, err := from.Call(to.EndpointForPort(80), apps.AppCallOptions{})
+
+				for _, r := range resp {
+					for _, h := range hosts {
+						if strings.HasPrefix(r.Hostname, h+"-") {
+							hitCount[id][h]++
```

Comment:
This is racy, you'll need to either use sync.Map or locks to protect the shared data structures accessed from multiple go-routines.

Reply:
I was struggled with the concurrency then. Since each goroutine accesses different map in the slice, do I still need to add lock to protect them?